Decouple Publicize Connection Lists from UI Take 2

mdawaffe opened the pull request:
Take 2 of the excellent #9955 by @c-shultz. See #9039.

Updated to work on top of #10043.

#### Changes proposed in this Pull Request:
* Adds a method to `Publicize_Base` to get the list of Publicize Connections suitable for use in UI.
* Cleans up logic v. presentation boundary in `Publicize_UI`.

#### Testing instructions:
##### Automated Tests
1. `yarn docker:phpunit --filter=WP_Test_Publicize`

##### Manual Tests
1. Under an *admin* account, set up a test site with two publicize accounts: one "global" (available to all users of the site), one not global.
2. As that *admin*, go to the Classic (non-Gutenberg) editor, and look at the Publicize UI for a new post.
3. You should see two checked checkboxes: one for each connection. You should be able to uncheck each (the inputs should not be disabled).
4. Publish the post.
5. You should see the publicized content (tweet, FB update, etc.) on both connections.
6. Under an *author* account, go to the Classic (non-Gutenberg) editor, and look at the Publicize UI or a new post.
7. You should see one checked-but disabled checkbox for the global connection. You should not see any evidence at all of the admin's non-global connection.
8. Publish a post as that *author*.
9. See the publicized content (tweet, FB update, etc.) on the global connection.

#### Proposed changelog entry for your changes:
None? Part of a larger initiative to provide Publicize data via REST API.

mdawaffe committed:
Publicize: Decouple Connection List from UI

Based on #9955 by @c-shultz.

"Rebased" on top of #10043

Moves Connection List Data "getter" out of `Publicize_UI` and in to
`Publicize_Base`.

mdawaffe committed:
Publicize: The comma separated list of connections should be separated by commas :)

The JS does this. So should the PHP.

mdawaffe committed:
Publicize: Tweak Tests

They fail :) But at least they now look at the right data
since the data structures changd from #9955 to b17be1f8929a40c3de227a5455510eafa544db79.

mdawaffe committed:
Publicize: Bump `@since`

mdawaffe committed:
Publicize: Move `->done_sharing_post()` out of `Publicize_Base`

It has Jetpack implementation details.

mdawaffe committed:
Publicize: use correct variable :)

mdawaffe committed:
Publicize: Remove the `visible` property.

Disabled checkboxes are never included in form submissions.

This is why we need the hidden form fields whenever we have a checked,
disabled checkbox.

Always show the checkbox (in the correct checked state). Look at
`toggleable` and `enabled` to see if we need the hidden form field.

mdawaffe committed:
Publicize: Fix tests

Be explicit about expected global/normal behavior.

ockham commented on modules/publicize/publicize.php:
```diff
@@ -510,6 +510,215 @@ function test_publicize_conns() {
+				$connection_list[] = array(
+					'unique_id'    => $unique_id,
+					'service_name' => $service_name,
+					'label'        => $label,
```
#9955 used to also provide [`displayName`](https://github.com/Automattic/jetpack/pull/9955/files#diff-2619864631f3665a877cdd9148da0072R594) (i.e. `label` without service label). This came in handy for Calypso, where we indicate the service name through icons:

![bildschirmfoto von 2018-10-25 21-05-06](https://user-images.githubusercontent.com/96308/47524263-68d2aa80-d89a-11e8-88dd-cec4cff88314.png)

Can we add back `display_name`, or even ditch `label` (and move concatenation with the service label to `ui.php`)?

ockham replied:
Here's a [workaround](https://github.com/Automattic/wp-calypso/pull/28102/commits/073f9561df6122855676fd5a60286fe3e8c29dca), but I'd rather not :)

mdawaffe replied:
Yes! I'll get this sorted shortly.

mdawaffe replied:
Done in 3995e0850

mdawaffe replied:
(And tests fixed in c4f653cbe :))

ockham replied:
Awesome, thanks!

mdawaffe committed:
Publicize: Output Service Label and Connection Username separately.

Output both these human-readable properties separately so that clients can
decide for themselves how to display them.

mdawaffe committed:
Publicize: Fix the tests I broke in 3995e08502f7a02493e46d154fa73b075b006bc0

ockham commented on modules/publicize/publicize.php:
```diff
@@ -510,6 +510,210 @@ function test_publicize_conns() {
+
+
+				// Was this connection (OR, old-format service) already Publicized to?
+				$done = ! empty( $post ) && (
```
Can we drop this check, and use `$post_id` instead of `$post->ID` below ([L569](https://github.com/Automattic/jetpack/pull/10396/files#diff-2f5f58835a08193f26adccabf4ae5ef3R567), [R569](https://github.com/Automattic/jetpack/pull/10396/files#diff-2f5f58835a08193f26adccabf4ae5ef3R569)) ? (Because of the check on [R544](https://github.com/Automattic/jetpack/pull/10396/files#diff-2f5f58835a08193f26adccabf4ae5ef3R544) and below) Or will `get_post_meta()` choke on a `null` post ID?

ockham commented on modules/publicize/publicize.php:
```diff
@@ -510,6 +510,210 @@ function test_publicize_conns() {
+							get_post_meta( $post_id, $this->POST_SKIP . $unique_id, true )
+							||
+							// Old flags
+							get_post_meta( $post->ID, $this->POST_SKIP . $service_name )
```
See https://github.com/Automattic/jetpack/pull/10396/files#r228333272 -- here it's even a bit stranger since [R598](https://github.com/Automattic/jetpack/pull/10396/files#diff-2f5f58835a08193f26adccabf4ae5ef3R598) is actually using `$post_id` already...

ockham reviewed (dismissed):
Left a few comments -- nothing major tho.

I tested this with the same instructions as #10043. LGTM! Ready to :ship: IMO

mdawaffe committed:
Publicize: Variable Consistency

Props @ockham.

jeherve reviewed (dismissed):
I ran some tests with the different interfaces, with different Publicize connections, and with different statuses in wp-admin; I could not spot any issues. I think you can merge when ready.

mdawaffe committed:
Publicize: Rename `done_sharing_post()` to `is_post_done_sharing()`

mdawaffe committed:
Publicize: Rename `is_post_done_sharing()` to `post_is_done_sharing()`

Consistency with other `post_is...()` functions.

oskosk reviewed (approved):
LGTM! thanks for the instructions!

dereksmart reviewed (approved):
Functionality tested and looked at the two commits since last review.
